src/rpc/client.rs

```diff
@@ -258,193 +118,259 @@ impl RpcClient {
+    fn regions_for_range(
+        self: Arc<Self>,
         range: BoundRange,
-        cf: Option<ColumnFamily>,
-    ) -> impl Future<Output = Result<()>> {
-        let scan: ScanRegionsContext<(), Option<ColumnFamily>> = ScanRegionsContext::new(range, cf);
-        let inner = Arc::clone(&self.inner);
-        loop_fn((inner, scan), |(inner, scan)| {
-            inner.locate_key(scan.start_key()).and_then(|location| {
-                let region = location;
-                let cf = scan.state.clone();
-                inner
-                    .clone()
-                    .region_context_by_id(region.id())
-                    .map_ok(|(region, client)| {
-                        (scan, region.range(), RawContext::new(region, client, cf))
-                    })
-                    .and_then(|(scan, region_range, context)| {
-                        let (ref start_key, ref end_key) = scan.range;
-                        let start_key = start_key.clone();
-                        let end_key = end_key.clone().expect("end key must be specified");
-                        context
-                            .client()
-                            .raw_delete_range(context, start_key, end_key)
-                            .map_ok(|_| (scan, region_range))
-                    })
-                    .map_ok(|(mut scan, region_range)| match scan.next(region_range) {
-                        ScanRegionsStatus::Continue => Loop::Continue((inner, scan)),
-                        ScanRegionsStatus::Break => Loop::Break(()),
-                    })
-            })
+    ) -> impl Stream<Item = Result<RawContext>> {
+        let (start_key, end_key) = range.into_keys();
+        stream_fn(Some(start_key), move |start_key| {
+            let start_key = match start_key {
+                None => return Either::Right(ready(Ok(None))),
+                Some(sk) => sk,
+            };
+            let end_key = end_key.clone();
+            let this = self.clone();
+            Either::Left(
+                self.clone()
+                    .get_region(&start_key)
+                    .and_then(move |location| {
+                        this.raw_from_id(location.id()).map_ok(move |context| {
+                            let region_end = context.region.end_key();
+                            if end_key.map(|x| x < region_end).unwrap_or(false)
+                                || region_end.is_empty()
+                            {
+                                return Some((None, context));
+                            }
+                            Some((Some(region_end), context))
+                        })
+                    }),
+            )
+        })
+    }
+
+    fn group_tasks_by_region<Task>(
+        self: Arc<Self>,
+        tasks: Vec<Task>,
+    ) -> impl Stream<Item = Result<(RegionId, Vec<Task>)>>
+    where
+        Task: GroupingTask,
+    {
+        let tasks: VecDeque<Task> = tasks.into();
```

Comment:
Is this efficient? It *looks* expensive.

Reply:
It is very cheap, it just copies the pointer (yay for ownership!) as long as the capacity of `tasks` is a power of two (which it should be if it has not been created using `with_capacity`.